### PR TITLE
feat: add update-stripe-card route

### PR DESCRIPTION
**Pull request by ahmaxed**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #XXXXX

### CODE REVIEW

File: client/i18n/locales/english/translations.json [hunk 1]
```diff
@@ -377,6 +378,10 @@
     "reset-warn-2": "This cannot be undone",
     "scrimba-tip": "Tip: If the mini-browser is covering the code, click and drag to move it. Also, feel free to stop and edit the code in the video at any time.",
     "chal-preview": "Challenge Preview",
+    "donation-record-not-found": "Your donation record has not been found.",
+    "sign-in-card-update": "Sign in to update your card",
+    "card-has-been-updated": "Your card has been updated successfully.",
+    "contact-support-mistake": "If you think there has been a mistake, please contact support at donors@freecodecamp.org",
```
Comment (naomi-lgbt) [inline client/i18n/locales/english/translations.json:384]: Given that we have a support email, and don't want general support requests erroneously sent to donors inbox, can we:

```suggestion
    "contact-support-mistake": "If you think there has been a mistake, please contact us at donors@freecodecamp.org",
```

Comment (ahmaxed) [reply client/i18n/locales/english/translations.json:384]: Good point

File: api-server/src/server/utils/donation.js [hunk 2]
```diff
@@ -314,3 +315,43 @@ export async function createStripeCardDonation(req, res, stripe) {
...
+    const donation = await Donation.findOne({
+      where: { userId: ObjectId(id) }
+    });
+
+    if (!donation) throw Error();
+
+    const { customerId, subscriptionId } = donation;
+
+    // updating customer payment method is handled by webhook handler
+    const session = await stripe.checkout.sessions.create({
```
Comment (ojeytonwilliams) [inline api-server/src/server/utils/donation.js:339]: Sorry for the stupid question, but does this return the same `subscription` id as we send it? I just want to be 100% sure that updating a subscription doesn't change the id!

File: e2e/update-stripe-card-default.spec.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
+import { test, expect, type Page } from '@playwright/test';
+import translations from '../client/i18n/locales/english/translations.json';
+
+let page: Page;
+test.beforeAll(async ({ browser }) => {
+  page = await browser.newPage();
+  await page.goto('/update-stripe-card');
+});
+
```
Comment (ojeytonwilliams) [inline e2e/update-stripe-card-default.spec.ts:9]:
```suggestion
```
Sharing pages between tests is a bit of an anti-pattern. Instead it's fine to simply write a larger test.

File: e2e/update-stripe-card-default.spec.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
+import { test, expect, type Page } from '@playwright/test';
```
Comment (ojeytonwilliams) [inline e2e/update-stripe-card-default.spec.ts:1]:
```suggestion
import { test, expect } from '@playwright/test';
```

File: e2e/update-stripe-card-default.spec.ts [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    await expect(page).toHaveTitle(
+      `${translations.misc['update-your-card']} | freeCodeCamp.org`
+    );
+  });
+  test('should display the info for signedin non-donor', async () => {
+    const h1 = page.locator('h1');
+    await expect(h1).toHaveText(
+      `${translations.learn['donation-record-not-found']}`
+    );
+  });
```
Comment (ojeytonwilliams) [inline e2e/update-stripe-card-default.spec.ts:22]:
```suggestion
  test('should render correctly', async ({ page }) => {
    await expect(page).toHaveTitle(
      `${translations.misc['update-your-card']} | freeCodeCamp.org`
    );
    const h1 = page.locator('h1');
    await expect(h1).toHaveText(
      `${translations.learn['donation-record-not-found']}`
    );
  });
```

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]

File: client/src/utils/ajax.ts [hunk 6]
```diff
@@ -238,6 +238,10 @@ export function addDonation(body: Donation): Promise<ResponseWithData<void>> {
   return post('/donate/add-donation', body);
 }
 
+export function updateStripeCard() {
+  return post('/donate/update-stripe-card');
```
[the remark this answers is not in the document]
Comment (ojeytonwilliams) [reply client/src/utils/ajax.ts:242]: The server would need updating, too, sorry!

File: api-server/src/server/utils/donation.js (hunk 2 above)
Comment (ahmaxed) [reply api-server/src/server/utils/donation.js:339]: So this just creates the checkout session and the returned object does not include a subscription id. In the cloud function, we are updating the user's default_payment_method and the subscription is not affected in any way.

Comment (ojeytonwilliams) [reply api-server/src/server/utils/donation.js]: Great, thanks for clarifying.

**Review by ojeytonwilliams — APPROVED** [review]
The code looks good, thanks. I haven't tested it out, though. Unless someone else has time to do that before merging, we should make 100% sure we check this on staging.

**Comment by ahmaxed** [comment]
Definitely, This page is not linked anywhere. It will be used once we add this route to the payment provider's settings.
We should be able to test this page even on production before it gets used.

**Review by naomi-lgbt — APPROVED** [review]
Think this should be fine to land. Not seeing any glaring issues.

**Pull request merged**